src/main/kotlin/org/rust/lang/core/lexer/LexerBaseEx.kt

```diff
@@ -0,0 +1,75 @@
+package org.rust.lang.core.lexer
+
+import com.intellij.lexer.LexerBase
+import com.intellij.psi.tree.IElementType
+
+/**
+ * Small utility class to ease implementing [LexerBase].
+ */
+abstract class LexerBaseEx : LexerBase() {
+    private var state: Int = 0
+    private var tokenStart: Int = 0
+    private var tokenEnd: Int = 0
+    private lateinit var bufferSequence: CharSequence
+    private var bufferEnd: Int = 0
+    private var myTokenType: Lazy<IElementType?> = lazyOf(null)
+
+    /**
+     * Determine type of the current token (the one delimited by [tokenStart] and [tokenEnd]).
+     */
+    protected abstract fun determineTokenType(): IElementType?
+
+    /**
+     * Find next token location (the one starting with [tokenEnd] and ending somewhere).
+     * @return end offset of the next token
+     */
+    protected abstract fun locateToken(start: Int): Int
+
+    override fun start(buffer: CharSequence, startOffset: Int, endOffset: Int, initialState: Int) {
+        bufferSequence = buffer
+        bufferEnd = endOffset
+        state = initialState
+
+        tokenStart = startOffset
+        tokenEnd = locateToken(tokenStart)
+        myTokenType = lazy { determineTokenType() }
+    }
+
+    override fun advance() {
+        tokenStart = tokenEnd
+        tokenEnd = locateToken(tokenStart)
+        myTokenType = lazy { determineTokenType() }
```

Comment:
Can this be changed to `myTokenType = determineTokenType()`? This might be less efficient (but I guess the token type is queried after each advance anyway), but it is much simpler and is probably more efficient (lazys are synchronized).